docs/AUTH.md

```diff
@@ -1,36 +1,387 @@
++ "."
+  signature
+```
+
+The sample encoded JWT (can be viewed on [jwt.io](https://jwt.io)):
+
+```text
+eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJhcHBJZCI6IndlYi1jbGllbnQtYXBwLWlkIiwiYXV0aG9yIjoibXktdXNlci1pZCIsInVybSI6eyJ4eXotaHViIjp7InJlYWRGZWF0dXJlcyI6W3sic3RvcmFnZUlkIjoiZGV2LSoifV19fSwiaWF0IjoxNzA0MDYzNTk5LCJleHAiOjE3MDQwNjcxOTl9.g7maKIpDQ6d8MoC7lPQDa_6BLKV5HhpN9t1BkcdFmNSetc-dHIcor_mhvc4GNpJELEMCBiTiF8RdlY_PEOooJc4Ixx5yWFoeIEaKv-aunvf6TZsOlD8F5KX8CmL8QzEO7t8YrSVz-F3WYrw1rmnl_1WC2tscMUBvfFHRifq3h7F46ZMswO6fm8AGHW0bbSeDCwK2VcjkYOwGVYWmSPodtxT7ie8uxJlAFxaCGzxV1WkVnrqIZFdPcnq3hM_FjbSw01MxOD3qdiL47HRXQnvOzsKjhi5ihClihwiua4N9xOeq2I8nX5_2YJIRWjS8pAozRp7cfnhb15Sm8JevqEwz1A
+```
+
+JWT payload is expected to have custom claims `appId`, `author` and `urm`, which is then populated into [NakshaContext](../here-naksha-lib-core/src/main/java/com/here/naksha/lib/core/NakshaContext.java) whenever REST API request is picked up for processing.
+
+```json
+{
+    "appId": "web-client-app-id",
+    "author": "my-user-id",
+    "urm": {
+        "xyz-hub": {
+            "readFeatures": [
+                {
+                    "storageId": "dev-*"
+                }
+            ]
+        }
+    },
+    "iat": 1704063599,
+    "exp": 1704067199
+}
+```
+
+### Auth Modes
+
+Service can be executed in two modes based on [AuthorizationType](../here-naksha-app-service/src/main/java/com/here/naksha/app/service/http/auth/Authorization.java) specified as part of startup [config](../here-naksha-lib-hub/src/main/java/com/here/naksha/lib/hub/NakshaHubConfig.java):
+
+* `DUMMY`
+  - Dummy mode
+  - useful, when we want to run service in local / test / dev environment, where security is not that important
+  - it will use internally generated super-user URM as part of NakshaContext to allow full access to all resources
+* `JWT`
+  - Real JWT mode
+  - useful, for cloud / prod environment, where security is MUST
+  - it will validate the JWT as part of each REST API request and extract the URM for further authorization checks
+  - Absence of or Invalid JWT will result into Http error code 401 - Unauthorized
+
+
+
+
+---
+
+## 3. Supported Resource Actions and Attributes
+
+Matrix of all supported **Actions** and **Attributes** for validating authorization against individual resource operation:
+
+**NOTE** for **Attributes**:
+  * For all, **exact** String value comparison is supported by default.
+  * For some, **wild-card** value (e.g. `storage-dev*`) is also supported (explicitly mentioned where applicable)
+  * For all, **List** of values is supported by default.
+
+**NOTE** for **Actions**:
+  * **Limited View** - means, read of resource is allowed BUT without exposing `properies` object (so typically one can read `id`, `title`, `description` etc but not `properties` object)
+  * **Filtered** read restricted resource instances - means, instances that are read restricted, they will be filtered from the result rather than raising exception like forbidden access
+
```

Comment:
No exceptions, the authorization matrices are a general purpose tool. It should not have exceptions. Generally I propose:
- The ARM is iterated and compared against the URM by service and action.
- Every entry in the ARM must have at least one matching entry (being greater or equal) in the URM
- The ARM must always be complete, it must not have missing attributes and must not contain partial values
- When comparing an attribute from the ARM against the URM, the best match is used. This means the URM attribute that is as closest. So if there is a URM attributes "foo*" and "foo_a*" and "foo_a_b" and the ARM attribute name is "foo_a_c", then "foo_a*" is selected.
- When comparing the value of the ARM against the URM, the URM value defines the operation.

The values in the ARM allowed are only:
- String
- Boolean
- Double

The operations in the URM are:
- _null_ = The value is _null_
- "string" = The ARM value must be the same string
- 12 = The ARM value must be double and of the exact same value
- "value*" = The ARM value must be a string, starting with the given value before the "*", if the value should end with a star, the last star need to be duplicated, so "value**" means exact match for "value*"
- {} = The ARM value can be any value

If the value in the URM is an object, this is a query object. Currently we do only define the behavior for the "match-any" query, which is an empty object. In the future, we may expand this using something for is-greater-than and alike.

Reply:
about the tags, after having a discussion with Andi about some UPM rights, I would suggest to change the way how tags are added. I think, the best is to define:

- In the ARM, the value can be an array of values, meaning, that multiple values are available
- In the ARM, the value can be an object of values, meaning, that multiple key-value pairs are available

in the URM we define that for every attribute generally a query is added. there are some shortcuts for commonly used queries (already described above), additionally they can always be used with full qualified notation. we define the query as:
```json
{
  "op": "<operation>",
  "args": [arguments],
  // optional operation modifiers
}
```
For example, "eq" for equals, "ne" for not equals. For each operator we define how they act on single values, values-array and values-map. This allows us to introduce future operations for new use cases without the need to change the source of the Naksha service, which will always generate a single ARM attribute map for each resource (this means, exactly one ARM attribute map or each resource).

Even while this looks complicated, it is not really complicated to implement. We only iterate all attribute-maps in the ARM and then iterate all URM entries, getting a "checker" for each operation and then ask "Checkers.get(opId).matches(op, arm_attr)". Note that the op can be either a map or a single value (string, boolean, null or double).

Reply:
I played a bit around and this seems to be a feasible and good syntax (so array of op and arguments):
Example of URM:
```json
{
  "naksha": {
    "readFeatures": [
      {"id": "foo"},
      {"id": ["eq","foo"]}, // same as above
      {"id": "exa*"},
      {"id": ["startsWith","exa"]}, // same as above
      {
        "tags": {
          "isRoad":["exists"],
          "salary":["gte", 5000],
          "is*": ["eq", true]
        }
      }
    ]
  }
}
```
and ARM:
```json
{
  "readFeature": [
    {
      "id": "foo",
      "tags": {
        "isRoad": true,
        "salary": 5500
      }
    },
    {
      "id": "bar",
      "tags": {
        "isRoad": true,
        "salary": 4900
      }
    }
  ]
}
```

Reply:
These operations is not something that I could recollect was discussed earlier.
1. To me, this sounds to be going in a direction of "filtering" strategy than "authorization checks". Is that intended?
2. Can you brief what is the real use-case or problem we are trying to solve? Or it is seen as anticipation?

ONLY after answer for both above questions are positive, I have follow-up questions:

1. Is there a value implementing this support in current scope? OR can be split into two phases?
2. It seems you have already thought it well, so would you mind contributing via merge PR? If not, I will try addressing sometime later (but will be likely delayed).

Reply:
Authorization is to some degree filtering, especially when reading features. Assume we have map-feedback objects, they can have three tags being **ref_{upmId}**, **public**, **topology** and **highway** (lets for now ignore the UPM use-case, and that we internally have tags available as maps). Now assume, we want to grant access to all features that are in your ownership or being public. How can you do this, and how can the server verify this?

From URM side, you would need to say:
```json
{
  "tags": ["anyOf", "ref_${user.id}", "public"]
}
```
Actually, UPM is able to replace the **${user.id}** with the UPM-ID of the currently operating user (**author**). This would lead to a concrete URM:
```json
[
  {
    "tags": ["anyOf", "ref_faf23r23rd", "public"]
  }
]
```

Note, the replacement is already supported, but yet nor in arrays, only in plain strings.

Now, from the service side, it would receive this URM and generate for every feature that should be read (post-filtering), the following feature filter:

```json
[
  {
    "id": "${id}",
    "tags": ["ref_faf23r23rd", "public", "topology", "highway"],
    // ... other filter values
  }
]
```

Now, the service can directly call the lib-auth function to compare and get a result _true_ or _false_. I see no way to implement this with the current library, I thought about adding pseudo attributes like "tag.{name}" and alike, but all this would be just bad working workarounds that all miss some use cases.

In the beginning I would suggest the following for URM values:

- `"foo"`: ["eq", (string)]
- `"foo*"`: ["startsWith", (string)] 
- `5` or `5.0`: ["eq", (double)]
- `true` or `false`: ["eq", (boolean)]
- else: ["operationName", arguments]
  - Arguments must be **string**, **double** or **boolean**

For the ARM I would suggest the following:

- `"string"`: The attribute has a single value, being a **string**
- `5` or `5.0`: The attribute has a single value, being a **double**
- `true` or `false`: The attribute has a single value, being a **boolean**
- `[(value),...]`: The attribute has multiple values, being **string**, **double** or **boolean**
- `{key:value, ...}`: The attribute has multiple key-value pairs with keys being **string**, values being **string**, **double** or **boolean**

Actually, every operation from the URM defines what parameters it requires and what the result on the different target types is. Currently, we will support these operations:

- `isNull`: Tests if the value is _null_.
- `eq`,`ne`,`gt`,`gte`,`lt`,`lte`: Requires one argument, compares the argument against the ARM attribute value. Always returns _false_ when the ARM value is array or map.
- `in`: Tests if the value is in the given range. Requires two arguments, the lower bound and upper bounds, both inclusive.
- `startsWith`: Requires one argument being **string**. Matches if the ARM attribute value is as well a **string** and starts with the given argument string. In any other case, results in _false_.
- `anyOf`: Accepts multiple arguments and tests if one of the given arguments is contained in ARM attribute. Allows any ARM attribute value. For scalar values (**string**, **double** or **boolean**) a direct compare is done, for maps and arrays all values are tested and if at least one of arguments is in the map or array, it matches (nxm compare). Otherwise, it returns _false_.
- `allOf`: The same as `anyOf`, but it only matches if all arguments are in the attribute value. This means, if multiple arguments are given, scalar values (**string**, **double** or **boolean**) will never match.
- `oneOf`: The same as `anyOf`, but it only matches if exactly one argument matches the attribute values. It fails if multiple match or if none matches.
- `anyKeyOf`: Accepts multiple arguments that must all be **string**s. Tests if the attribute values is a map and if it is, if at least one of the given arguments is contained in the map.
- `allKeyOf`: Same as `anyKeyOf`, but only matches if all keys are in the ARM attribute map.
- `oneKeyOf`: Same as `anyKeyOf`, but only matches if exactly one argument exists as key in the ARM attribute map, fails if multiple or none is contained.

More methods can be added later, when needed.

Reply:
It covers one use case in UPM, plus is anticipation of some use cases that will come with Map Feedback and Geospace API rewrite. For UPM the current use case is, that we want to check, that a user can only read other users, being in the communities he himself is in. So the user has a communityIds attribute (being an array of ids) and the URM will have a anyOf, with the community-ids of the principal.